repo/content/content_manager_indexes.go

```diff
@@ -87,48 +89,52 @@ func (bm *Manager) compactAndDeleteIndexBlobs(ctx context.Context, indexBlobs []
 		return nil
 	}
 
-	formatLog(ctx).Debugf("compacting %v contents", len(indexBlobs))
+	formatLog(ctx).Debugf("compacting %v index blobs", len(indexBlobs))
 
-	t0 := time.Now() // allow:no-inject-time
 	bld := make(packIndexBuilder)
 
+	var inputs, outputs []blob.Metadata
+
 	for _, indexBlob := range indexBlobs {
 		if err := bm.addIndexBlobsToBuilder(ctx, bld, indexBlob, opt); err != nil {
-			return err
+			return errors.Wrap(err, "error adding index to builder")
 		}
+
+		inputs = append(inputs, indexBlob.Metadata)
 	}
 
 	var buf bytes.Buffer
 	if err := bld.Build(&buf); err != nil {
 		return errors.Wrap(err, "unable to build an index")
 	}
 
-	compactedIndexBlob, err := bm.writePackIndexesNew(ctx, buf.Bytes())
+	compactedIndexBlob, err := bm.indexBlobManager.writeIndexBlob(ctx, buf.Bytes())
 	if err != nil {
 		return errors.Wrap(err, "unable to write compacted indexes")
 	}
 
-	formatLog(ctx).Debugf("wrote compacted index (%v bytes) in %v", compactedIndexBlob, time.Since(t0)) // allow:no-inject-time
-
+	// compaction wrote index blob that's the same as one of the sources
+	// it must be a no-op.
```

Comment:
Sorry, I missed / didn't realize this earlier.
If the output is the same as one of the inputs, it does not necessarily mean that the operation is a no-op.
You could have a situation where the inputs =  `{n1, n2, n3}` and output = `{n1}`, when say `n2` and `n3` _"cancel"_ each other. Suppose `n2` only has a content creation for `Ci`, and `n3` only has a deletion for `Ci`, that is old enough to be discarded. And suppose `{n1}` only has an index entry for `Ck != Ci`, then the output of the compaction would be `{n1}`.
One simple way of dealing with this is removing `n1` from the inputs. Essentially, the compaction log is saying which _input_ indices are superseded. Also, this suggest that a better name for the field is "superseded" or something along those lines.

Reply:
Jarek showed another case were the output could correspond to the compaction of a _previous_ input. In that case, it is not safe to use the "mark as superseded" approach described above.